File: 15-Dynamic-Programming/10-Interval-DP/cpp/burst_balloons.cpp
/* Burst Balloons | LeetCode 312 | Tags: interval-dp */
#include <bits/stdc++.h>
using namespace std;
int maxCoins(vector<int>& nums){nums.insert(nums.begin(),1);nums.push_back(1);int n=nums.size();
vector<vector<int>>dp(n,vector<int>(n,0));
for(int len=2;len<n;len++)for(int i=0;i+len<n;i++){int j=i+len;for(int k=i+1;k<j;k++)dp[i][j]=max(dp[i][j],dp[i][k]+dp[k][j]+nums[i]*nums[k]*nums[j]);}return dp[0][n-1];}
int main(){vector<int>nums={3,1,5,8};cout<<maxCoins(nums)<<endl;} // 167

File: 15-Dynamic-Programming/10-Interval-DP/java/BurstBalloons.java
import java.util.*;
public class BurstBalloons { public static int maxCoins(int[] nums){int n=nums.length+2;int[] a=new int[n];a[0]=a[n-1]=1;System.arraycopy(nums,0,a,1,nums.length);
    int[][] dp=new int[n][n];for(int len=2;len<n;len++)for(int i=0;i+len<n;i++){int j=i+len;for(int k=i+1;k<j;k++)dp[i][j]=Math.max(dp[i][j],dp[i][k]+dp[k][j]+a[i]*a[k]*a[j]);}return dp[0][n-1];}
    public static void main(String[] args){System.out.println(maxCoins(new int[]{3,1,5,8}));}}
